File: bindings/python/doc/simplenet.py
from __future__ import print_function
import numpy as np
import cntk as C
from cntk.learners import sgd
from cntk.logging import ProgressPrinter
from cntk.layers import Dense, Sequential

def generate_random_data(sample_size, feature_dim, num_classes):
     # Create synthetic data using NumPy.
     Y = np.random.randint(size=(sample_size, 1), low=0, high=num_classes)

     # Make sure that the data is separable
     X = (np.random.randn(sample_size, feature_dim) + 3) * (Y + 1)
     X = X.astype(np.float32)
     # converting class 0 into the vector "1 0 0",
     # class 1 into vector "0 1 0", ...
     class_ind = [Y == class_number for class_number in range(num_classes)]
     Y = np.asarray(np.hstack(class_ind), dtype=np.float32)
     return X, Y

def ffnet():
    inputs = 2
    outputs = 2
    layers = 2
    hidden_dimension = 50

    # input variables denoting the features and label data
    features = C.input_variable((inputs), np.float32)
    label = C.input_variable((outputs), np.float32)

    # Instantiate the feedforward classification model
    my_model = Sequential ([
                    Dense(hidden_dimension, activation=C.sigmoid),
                    Dense(outputs)])
    z = my_model(features)

    ce = C.cross_entropy_with_softmax(z, label)
    pe = C.classification_error(z, label)

    # Instantiate the trainer object to drive the model training
    lr_per_minibatch = C.learning_parameter_schedule(0.125)
    progress_printer = ProgressPrinter(0)
    trainer = C.Trainer(z, (ce, pe), [sgd(z.parameters, lr=lr_per_minibatch)], [progress_printer])

    # Get minibatches of training data and perform model training
    minibatch_size = 25
    num_minibatches_to_train = 1024

    aggregate_loss = 0.0
    for i in range(num_minibatches_to_train):
        train_features, labels = generate_random_data(minibatch_size, inputs, outputs)
        # Specify the mapping of input variables in the model to actual minibatch data to be trained with
        trainer.train_minibatch({features : train_features, label : labels})
        sample_count = trainer.previous_minibatch_sample_count
        aggregate_loss += trainer.previous_minibatch_loss_average * sample_count

    last_avg_error = aggregate_loss / trainer.total_number_of_samples_seen

    test_features, test_labels = generate_random_data(minibatch_size, inputs, outputs)
    avg_error = trainer.test_minibatch({features : test_features, label : test_labels})
    print(' error rate on an unseen minibatch: {}'.format(avg_error))
    return last_avg_error, avg_error

np.random.seed(98052)
ffnet()

File: bindings/r/simplenet.R
library(cntk)
library(magrittr)

generate_data <- function(sample_size, feature_dim, num_classes) {
	# create synthetic data
	Y <- replicate(1, sample(0:num_classes, sample_size, rep = TRUE))
	
	# make sure the data is separable
	M <- matrix(rnorm(sample_size * feature_dim, 3), ncol = feature_dim)
	X <- M * as.vector(Y + 1)

	# one-hot encoding
	class_ind <- matrix(nrow = sample_size, ncol = num_classes)
	for (class in 1:num_classes) {
		class_ind[,class] <- as.numeric(Y == class)
	}
	Y <- class_ind

	list(X, Y)
}

inputs <- 2
outputs <- 2
layers <- 2
hidden_dimension <- 50

# input variables denoting features and label data
features <- input_variable(inputs)
label <- input_variable(outputs)

# instantiate the feedforward classification model
my_model <- sequential(
	layer_dense(hidden_dimension, activation = sigmoid),
	layer_dense(outputs)
)
z <- my_model(features)

ce <- cross_entropy_with_softmax(z, label)
pe <- metric_classification_error(z, label)

# instantiate trainer object
lr_per_minibatch <- learning_rate_schedule(0.125, UnitType$minibatch)
progress_printer <- ProgressPrinter()
optimizer <- sgd(z$parameters, lr = lr_per_minibatch)
trainer <- Trainer(z, c(ce, pe), c(optimizer), c(progress_printer))

# get minibatches of training data and train model
minibatch_size <- 25
num_minibatches <- 1024

aggregate_loss = 0L
for (i in 1:num_minibatches) {
	training_data <- generate_data(minibatch_size, inputs, outputs)

	# map input variable to minibatch data
	trainer %>% train_minibatch(c("features", "label"), training_data)
	#     trainer$train_minibatch(dict(features = train[1], label = train[2]))
	sample_count <- trainer$previous_minibatch_sample_count
	loss <- trainer$previous_minibatch_loss_average * sample_count
	aggregate_loss <- aggregate_loss + loss
}

last_avg_error <- aggregate_loss / trainer$total_number_of_samples_seen

test <- generate_data(minibatch_size, inputs, outputs)
error <- trainer$test_minibatch(dict(features = test[1], label = test[1]))
sprintf('Error rate on an unseen minibatch: %f',  error)
